### PR TITLE
[MRG] Two fixes to Categorical dimension of space

### PR DESCRIPTION
**Pull request by nfcampos**
- support 2 classes
  - LabelBinarizer outputs only 1 column for the 2 classes case (see https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/preprocessing/label.py#L423)
- support `None` as a category
  - useful for eg. class_weight of RandomForestClassifier
  - requires transforming None into/from a string value before/after passing/receiving it from LabelBinarizer

### CODE REVIEW

**Comment by betatim** [comment]
Thanks for the PR!

Could you add a test that would have caught the two class counting bug? So it doesn't come back 🐛

**Comment by nfcampos** [comment]
sure, do you want a test for `None` as a category as well?

File: skopt/space.py [hunk 1]
```diff
@@ -255,7 +258,8 @@ def rvs(self, n_samples=None, random_state=None):
 
     @property
     def transformed_size(self):
-        return len(self.categories)
+        l = len(self.categories)
+        return l if l != 2 else 1
```
Comment (betatim) [inline skopt/space.py]: `l` -> `size` maybe? Right now it looks a bit like `return 1 if 1 != 2 else 1` :)

**Comment by betatim** [comment]
Are there cases where you want to use `None` and numbers? It is much less clear in that case how to encode `None` but if someone has a bright idea we could do this in this PR. Not mandatory though.

File: skopt/space.py (hunk 1 above)
Comment (nfcampos) [inline skopt/space.py]: yep ill change it to size

File: skopt/space.py [hunk 2]
```diff
@@ -42,6 +42,9 @@ class _CategoricalEncoder:
     def __init__(self):
         """Convert labeled categories into one-hot encoded features."""
         self._lb = LabelBinarizer()
+        NONE_ = '__None__'
+        self._Xin = np.vectorize(lambda a: a if a is not None else NONE_)
+        self._Xout = np.vectorize(lambda a: a if a is not NONE_ else None)
```
Comment (betatim) [inline skopt/space.py]: rename it as `{en,de}code_none`? Something that is a bit more descriptive would be good.

**Comment by betatim** [comment]
> sure, do you want a test for `None` as a category as well?

All the tests all the time (if you are offering) :)

**Comment by nfcampos** [comment]
For optimizing hyper parameters of scikit classifiers actually what would be useful is a `Mixed` space that supports both numbers between `min` and `max` (like `Integer` or `Real`) as well as discrete categorical values (of which `None` could be one of them). but this would be in another PR

File: skopt/space.py (hunk 2 above)
Comment (MechCoder) [inline skopt/space.py]: Doesn't storing lambda as attributes have issues while pickling?

Comment (nfcampos) [inline skopt/space.py]: yep, that's better i'll rename

Comment (nfcampos) [inline skopt/space.py]: @MechCoder what do you suggest? is making it a regular function enough?

**Comment by MechCoder** [comment]
I believe you can do that with the Space API. no?
https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/tests/test_space.py#L129

**Comment by nfcampos** [comment]
in that line the catgeories and the numbers are for different parameters right?

**Comment by nfcampos** [comment]
I've addressed all feedback other than the `lambda` question, what do you guys suggest there?

**Comment by betatim** [comment]
The problem is that you [can't pickle a lambda](https://gist.github.com/betatim/ae42c092827d2f8e9d6e4f7219a8cadd). If you make it a method instead it will pickle. Or should it be a helper outside the class @MechCoder?

**Comment by nfcampos** [comment]
got it.
how about this?
```py
class CategoricalEncoder:
  @np.vectorize
  def encode_none(a):
    return a if a is not None else NONE_
```

File: skopt/space.py [hunk 3]
```diff
@@ -42,6 +42,9 @@ class _CategoricalEncoder:
     def __init__(self):
         """Convert labeled categories into one-hot encoded features."""
         self._lb = LabelBinarizer()
+        NONE_ = '__None__'
```
Comment (glouppe) [inline skopt/space.py]: Do we really need that variable? could be directly inside the lambda

File: skopt/space.py [hunk 4]
```diff
@@ -42,6 +42,9 @@ class _CategoricalEncoder:
...
+        self.encode_none = np.vectorize(lambda a: a if a is not None else NONE_)
+        self.decode_none = np.vectorize(lambda a: a if a != NONE_ else None)
```
Comment (glouppe) [inline skopt/space.py]: also there is not really a point in having those as attributes. Just define encode/decode  as regular functions, outside of the class

Comment (nfcampos) [inline skopt/space.py]: ok, I'll define them as functions outside

File: skopt/space.py [hunk 5]
```diff
@@ -255,7 +258,8 @@ def rvs(self, n_samples=None, random_state=None):
 
     @property
     def transformed_size(self):
-        return len(self.categories)
+        size = len(self.categories)
+        return size if size != 2 else 1
```
Comment (glouppe) [inline skopt/space.py]: I dont understand. Surely you are adding a new `None` category right, which be counted for.

Comment (nfcampos) [inline skopt/space.py]: this is a completely unrelated fix. see https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/preprocessing/label.py#L423 scikit's `LabelBinarizer` outputs arrays of `size == len(categories` for all cases other than passing in exactly two categories to `fit` in which case it outputs a single binary vector

Comment (glouppe) [inline skopt/space.py]: If my categories are `["a", "b"]` why wouldnt the size be 2?

Comment (nfcampos) [inline skopt/space.py]: no, it'll be 1, see this line https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/preprocessing/label.py#L441

Comment (nfcampos) [inline skopt/space.py]: and compare to the example just above with 4

Comment (glouppe) [inline skopt/space.py]: Okay thanks. Please add an inline comment then, this is really not obvious without knowing.

Comment (nfcampos) [inline skopt/space.py]: yeah good call, I'll add it

File: skopt/tests/test_space.py [hunk 6]
```diff
@@ -65,22 +65,40 @@ def test_integer():
...
 
+def test_categorical_transform_binary():
+    categories = ["apple", "orange"]
+    cat = Categorical(categories)
+
+    apple = [0.]
+    orange = [1.]
+
+    assert_equal(cat.transform(["apple"]).size, cat.transformed_size)
+    assert_array_equal(cat.transform(categories), [apple, orange])
```
Comment (MechCoder) [inline skopt/tests/test_space.py]: We should change `Categorical` to handle this special case, such that this outputs `[[1.0, 0.0], [0.0, 1.0]]`. The reason `LabelBinarizer` handles it this way is because you can infer what the other column is given the output of `transform`.

Comment (nfcampos) [inline skopt/tests/test_space.py]: why would we purposefully take up more memory when this is completely transparent to the user, is anyone supposed to directly use the `warped` space?

**Comment by MechCoder** [comment]
OK. So there are two unrelated changes:

1. The first one is the behaviour of Categorical in the case of the two-category case, for example ["a", b"]. It is correct to return either `[[1.0, 0.0], [0.0, 1.0]]` or `[[1.0], [0.0]]` (current). But I propose to change the output to `[[1.0, 0.0], [0.0, 1.0]]` such as over here (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/logistic.py#L648) to avoid further confusion as to why the number of dimensions are different,

2. The second one is the usage of `None` as a category which might require a bit more discussion. You can easily encode None to be a category and de-encode it back to None inside your function. 

Would you be able to split the changes, so that it is easier to review? Thanks!

**Comment by nfcampos** [comment]
sure

**Comment by nfcampos** [comment]
I've opened the two PRs with these 2 fixes, you can close this one
